File: external_apps/mobile_ios/mobile_ios/Core/Analytics/PerformanceMonitor.swift
//
//  PerformanceMonitor.swift
//  mobile_ios
//
//  Created for Analytics & Monitoring
//

import Foundation

struct PerformanceMetric {
    let name: String
    let durationMs: Int64
    let attributes: [String: String]
    let timestamp: Date
}

class PerformanceMonitor {
    static let shared = PerformanceMonitor()
    
    private var metrics: [PerformanceMetric] = []
    private var activeTraces: [String: Date] = [:]
    private let appStartTime = Date()
    
    private init() {}
    
    /// Start performance trace
    func startTrace(_ traceName: String) {
        activeTraces[traceName] = Date()
        print("PerformanceMonitor: Started trace: \(traceName)")
    }
    
    /// Stop performance trace
    func stopTrace(_ traceName: String, attributes: [String: String] = [:]) {
        guard let startTime = activeTraces.removeValue(forKey: traceName) else { return }
        
        let duration = Int64(Date().timeIntervalSince(startTime) * 1000)
        let metric = PerformanceMetric(
            name: traceName,
            durationMs: duration,
            attributes: attributes,
            timestamp: Date()
        )
        metrics.append(metric)
        
        print("PerformanceMonitor: Stopped trace: \(traceName) (\(duration)ms)")
        
        // TODO: Send to Firebase Performance
        // let trace = Performance.startTrace(name: traceName)
        // attributes.forEach { trace?.setValue($0.value, forAttribute: $0.key) }
        // trace?.stop()
    }
    
    /// Measure execution time of a block
    func measureTime<T>(_ traceName: String, block: () -> T) -> T {
        startTrace(traceName)
        defer { stopTrace(traceName) }
        return block()
    }
    
    /// Track network request performance
    func trackNetworkRequest(
        url: String,
        method: String,
        statusCode: Int,
        durationMs: Int64,
        requestSize: Int64 = 0,
        responseSize: Int64 = 0
    ) {
        let metric = PerformanceMetric(
            name: "network_request",
            durationMs: durationMs,
            attributes: [
                "url": url,
                "method": method,
                "status_code": "\(statusCode)",
                "request_size": "\(requestSize)",
                "response_size": "\(responseSize)"
            ],
            timestamp: Date()
        )
        metrics.append(metric)
        
        print("PerformanceMonitor: Network: \(method) \(url) - \(statusCode) (\(durationMs)ms)")
    }
    
    /// Track screen rendering time
    func trackScreenLoad(_ screenName: String, durationMs: Int64) {
        let metric = PerformanceMetric(
            name: "screen_load",
            durationMs: durationMs,
            attributes: ["screen_name": screenName],
            timestamp: Date()
        )
        metrics.append(metric)
        
        print("PerformanceMonitor: Screen load: \(screenName) (\(durationMs)ms)")
    }
    
    /// Get app start time
    func getAppStartTime() -> TimeInterval {
        return Date().timeIntervalSince(appStartTime)
    }
    
    /// Get all metrics
    func getMetrics() -> [PerformanceMetric] {
        return metrics
    }
    
    /// Get metrics by name
    func getMetricsByName(_ name: String) -> [PerformanceMetric] {
        return metrics.filter { $0.name == name }
    }
    
    /// Get average duration for metric
    func getAverageDuration(_ name: String) -> Int64 {
        let filtered = getMetricsByName(name)
        guard !filtered.isEmpty else { return 0 }
        return filtered.reduce(0) { $0 + $1.durationMs } / Int64(filtered.count)
    }
    
    /// Clear metrics
    func clearMetrics() {
        metrics.removeAll()
    }
}

File: external_apps/mobile_android/app/src/main/java/com/example/mobile_android/data/analytics/PerformanceMonitor.kt
package com.example.mobile_android.data.analytics

import android.content.Context
import android.os.SystemClock
import android.util.Log
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import javax.inject.Inject
import javax.inject.Singleton

data class PerformanceMetric(
    val name: String,
    val durationMs: Long,
    val attributes: Map<String, String> = emptyMap(),
    val timestamp: Long = System.currentTimeMillis()
)

@Singleton
class PerformanceMonitor @Inject constructor(
    @ApplicationContext private val context: Context
) {
    
    private val metrics = mutableListOf<PerformanceMetric>()
    private val activeTraces = mutableMapOf<String, Long>()
    
    private val _appStartTime = SystemClock.elapsedRealtime()
    
    /**
     * Start performance trace
     */
    fun startTrace(traceName: String) {
        activeTraces[traceName] = SystemClock.elapsedRealtime()
        Log.d("PerformanceMonitor", "Started trace: $traceName")
    }
    
    /**
     * Stop performance trace
     */
    fun stopTrace(traceName: String, attributes: Map<String, String> = emptyMap()) {
        val startTime = activeTraces.remove(traceName)
        if (startTime != null) {
            val duration = SystemClock.elapsedRealtime() - startTime
            val metric = PerformanceMetric(traceName, duration, attributes)
            metrics.add(metric)
            
            Log.d("PerformanceMonitor", "Stopped trace: $traceName (${duration}ms)")
            
            // TODO: Send to Firebase Performance
            // val trace = FirebasePerformance.getInstance().newTrace(traceName)
            // attributes.forEach { (key, value) -> trace.putAttribute(key, value) }
            // trace.stop()
        }
    }
    
    /**
     * Measure execution time of a block
     */
    inline fun <T> measureTime(traceName: String, block: () -> T): T {
        startTrace(traceName)
        return try {
            block()
        } finally {
            stopTrace(traceName)
        }
    }
    
    /**
     * Track network request performance
     */
    fun trackNetworkRequest(
        url: String,
        method: String,
        statusCode: Int,
        durationMs: Long,
        requestSize: Long = 0,
        responseSize: Long = 0
    ) {
        val metric = PerformanceMetric(
            name = "network_request",
            durationMs = durationMs,
            attributes = mapOf(
                "url" to url,
                "method" to method,
                "status_code" to statusCode.toString(),
                "request_size" to requestSize.toString(),
                "response_size" to responseSize.toString()
            )
        )
        metrics.add(metric)
        
        Log.d("PerformanceMonitor", "Network: $method $url - ${statusCode} (${durationMs}ms)")
    }
    
    /**
     * Track screen rendering time
     */
    fun trackScreenLoad(screenName: String, durationMs: Long) {
        val metric = PerformanceMetric(
            name = "screen_load",
            durationMs = durationMs,
            attributes = mapOf("screen_name" to screenName)
        )
        metrics.add(metric)
        
        Log.d("PerformanceMonitor", "Screen load: $screenName (${durationMs}ms)")
    }
    
    /**
     * Get app start time
     */
    fun getAppStartTime(): Long {
        return SystemClock.elapsedRealtime() - _appStartTime
    }
    
    /**
     * Get all metrics
     */
    fun getMetrics(): List<PerformanceMetric> = metrics.toList()
    
    /**
     * Get metrics by name
     */
    fun getMetricsByName(name: String): List<PerformanceMetric> {
        return metrics.filter { it.name == name }
    }
    
    /**
     * Get average duration for metric
     */
    fun getAverageDuration(name: String): Long {
        val filtered = getMetricsByName(name)
        return if (filtered.isNotEmpty()) {
            filtered.sumOf { it.durationMs } / filtered.size
        } else {
            0
        }
    }
    
    /**
     * Clear metrics
     */
    fun clearMetrics() {
        metrics.clear()
    }
}
